pkg/kv/store_message.go

```diff
@@ -14,54 +13,46 @@ type StoreMessage struct {
 }
 
 func (s *StoreMessage) GetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) error {
-	val, err := s.Store.Get(ctx, []byte(path))
+	v, err := s.Store.Get(ctx, []byte(path))
 	if err != nil {
-		return fmt.Errorf("failed on Get (path: %s): %w", path, err)
+		return err
 	}
-	return proto.Unmarshal(val, msg)
+	return proto.Unmarshal(v, msg)
 }
 
-func (s *StoreMessage) SetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) error {
-	val, err := proto.Marshal(msg)
+func (s *StoreMessage) GetMsgPredicate(ctx context.Context, path string, msg protoreflect.ProtoMessage) (Predicate, error) {
+	vp, err := s.Store.GetValuePredicate(ctx, []byte(path))
 	if err != nil {
-		return fmt.Errorf("failed on Set (path: %s): %w", path, err)
+		return nil, err
 	}
-	return s.Store.Set(ctx, []byte(path), val)
+	if msg == nil {
+		return vp.Predicate, nil
+	}
+	err = proto.Unmarshal(vp.Value, msg)
+	if err != nil {
+		return nil, err
+	}
+	return vp.Predicate, nil
 }
 
-func (s *StoreMessage) SetIf(ctx context.Context, path string, msg protoreflect.ProtoMessage, pred protoreflect.ProtoMessage) error {
+func (s *StoreMessage) SetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) error {
 	val, err := proto.Marshal(msg)
 	if err != nil {
-		return fmt.Errorf("failed on Marshal (path: %s): %w", path, err)
-	}
-	if pred == nil {
-		return s.Store.SetIf(ctx, []byte(path), val, nil)
+		return err
 	}
+	return s.Store.Set(ctx, []byte(path), val)
+}
 
-	// Calling Get here is suboptimial and ideally we would save the round trip to the store.
-	// Since Proto Marshaling can be inconsistent we must read and parse it first to have a consistent predicate compare.
-	// A future optimization here would be adding an ID to the Store interface, or a consistent marshaling protocol.
-	curr, err := s.Store.Get(ctx, []byte(path))
-	if err != nil {
-		return fmt.Errorf("failed on Get. Path (%s): %w", path, err)
-	}
-	currMsg := msg.ProtoReflect().New().Interface()
-	err = proto.Unmarshal(curr, currMsg)
+func (s *StoreMessage) SetMsgIf(ctx context.Context, path string, msg protoreflect.ProtoMessage, predicate Predicate) error {
+	val, err := proto.Marshal(msg)
 	if err != nil {
-		return fmt.Errorf("failed on Unmarshal (path: %s) %w", path, err)
-	}
-	if !proto.Equal(pred, currMsg) {
-		return fmt.Errorf("failed on predicate. Path (%s): %w", path, ErrPredicateFailed)
+		return err
 	}
-	return s.Store.SetIf(ctx, []byte(path), val, curr)
+	return s.Store.SetIf(ctx, []byte(path), val, predicate)
 }
 
 func (s *StoreMessage) Delete(ctx context.Context, path string) error {
-	err := s.Store.Delete(ctx, []byte(path))
-	if err != nil {
-		err = fmt.Errorf("failed on Delete (path: %s): %w", path, err)
-	}
-	return err
+	return s.Store.Delete(ctx, []byte(path))
```

Comment:
The error was here for a reason - this changes the multiparts tracker current behavior

Reply:
The comment was regarding the removal of the check in delete to see if the key exists and if not return an error

Reply:
The previous code - for any error in delete enrich the error with path - no extra verification at this level.
```go
        err := s.Store.Delete(ctx, []byte(path))
	if err != nil {
		err = fmt.Errorf("failed on Delete (path: %s): %w", path, err)
	}
	return err
```

Reply:
Sorry, my bad - it's handled somewhere else